### ISSUE TITLE
question for Nix integration

### ISSUE THREAD

**Issue by roberth** [pre-commit/pre-commit#1209]
Hi,

We've built a nix [integration](https://github.com/hercules-ci/nix-pre-commit-hooks) around pre-commit. (around, instead of inside pre-commit, as discussed later). It gives us reproducible, locked dependencies and a build cache among other things.
The way it works is that we create a symlink from the project to a unique path in the Nix store, which we also rely on for garbage collection.
This turned out to be a bit harder to integrate than we hoped, because we had an issue with pre-commit's own cache and we could not find a good way to either use it well or to bypass it. We ended up adding [this patch](https://github.com/hercules-ci/nix-pre-commit-hooks/blob/94e021bcd3bb2cd4949cdd6bcb41a907dc6033ba/nix/pre-commit-resolve-symlink.patch).

The patch still uses the cache, which we have to invalidate separately, because the behavior of `_non_cloned_repository_hooks` was too different from `_cloned_repository_hooks`. The non-cloned variation returned errors about 'language' not being defined in some hooks. I haven't found the time to investigate why this differs yet.

So I was wondering whether you'd welcome a change like this?

-----

*Why not integrate directly into pre-commit (yet?)*

I think it'd be cool if `pre-commit` supported Nix as a `language`, but I recommend not to build such a feature just yet. The main reason is that the Nix landscape is about to change because of a proposal for package definitions called "flakes". It seems this will have quite some impact. Also currently it's best to build all Nix hooks in one go, which may or may not be easy to implement.

**Comment by asottile** [comment]
the patch makes me think you're missing the point of `~/.cache/pre-commit` and `PRE_COMMIT_HOME` -- I'm curious what problems led you to that "solution"?

**Comment by asottile** [comment]
I'm confused why you need to subvert the internals, expand more on the problems requiring `clean`?

**Comment by roberth** [comment]
Essentially we had something like this:

```
layout after installation with nix-pre-commit-hooks:

|- project-a
|   |- .git
|   |- .pre-commit-config.yaml  (with a local repo reference to .pre-commit-hooks)
|   `- .pre-commit-hooks  ->  /nix/store/abcdef
`- project-b
    |- .git
    |- .pre-commit-config.yaml  (with a local repo reference to .pre-commit-hooks)
    `- .pre-commit-hooks  ->  /nix/store/654321

$ cd project-a
$ pre-commit run -a
... as expected ...

$ cd ../project-b
$ pre-commit run -a
... hooks from project-a are run! ...

$ ls ~/.cache/pre-commit/
db.db  README  repo3j57qgzy

```

**Comment by asottile** [comment]
yes but why, what doesn't work with vanilla

**Comment by roberth** [comment]
`pre-commit run -a` should not run hooks for `project-a` when in `project-b`

**Comment by asottile** [comment]
right, take a step back -- ignore all the nix stuff -- what doesn't work with vanilla pre-commit

**Comment by roberth** [comment]
tldr; thank you

I was unable to reproduce the [caching error](https://github.com/pre-commit/pre-commit/issues/1209#issuecomment-550369720) with a minimal example and `pre-commit` 1.18 or 1.20, so my original hypothesis about the cause of the problem may have been wrong.

So I'm afraid I have to conclude that pre-commit worked well and I made a mistake.

I don't know what went wrong, but the (lack of) design was a certainly a problem.
We can simplify our solution by generating only a `.pre-commit-config.yaml` containing local hooks.
This seems to make more sense, considering the docs:
> Repository-local hooks are useful when:
> - The scripts are tightly coupled to the repository and it makes sense to distribute the hook scripts with the repository.
> - Hooks require state that is only present in a built artifact of your repository (such as your app's virtualenv for pylint).
> - The official repository for a linter doesn't have the pre-commit metadata.

I did try this but I backtracked when I encountered [_has_unstaged_config](https://github.com/pre-commit/pre-commit/blob/1509dc497e77d742278ae9f6ed19481a24c972f2/pre_commit/commands/run.py#L243). That was wrong because that check does work as we'd like if the config is gitignored.

So there you have it. :duck: 
It seems like pre-commit will do the job just fine as-is in our - let's say, novel - use case.

Thank you for your patience and helping me figure this out.
Someone else might have been all grumpy about bending the project to some unusual requirements, but you weren't and I respect that a lot :)

**Comment by asottile** [comment]
:tada: neat!  good to hear you got to a solution!  fwiw, it's ~supposed to work with `.pre-commit-config.yaml` not-checked-in but it's possible it acts strangely when `.gitignore`d

**Comment by roberth** [comment]
Cool, thanks for confirming. But how can I expect it to act up when gitignored?

**Comment by asottile** [comment]
hmmmmm, that seems to work for me too:

```console
$ git init .
$ echo '/.pre-commit-config.yaml' > .gitignore
$ git add .gitignore
$ pre-commit  sample-config > .pre-commit-config.yaml
$ pre-commit install
pre-commit installed at .git/hooks/pre-commit
$ git status
On branch master

No commits yet

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)

	new file:   .gitignore

$ git commit -m "foo"
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check Yaml...........................................(no files to check)Skipped
Check for added large files..............................................Passed
[master (root-commit) c47d57e] foo
 1 file changed, 1 insertion(+)
 create mode 100644 .gitignore
```

**Comment by asottile** [comment]
when nix is more stable, perhaps it could also be added as a `language` as suggested above, I suspect it'll be pretty similar to `conda` (#1204)
